### ISSUE TITLE
Native CSS nesting support for css modules

### ISSUE THREAD

**Issue by equinusocio** [vercel/next.js#55053]
### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/magical-chatelet-9ddsys?file=%2Fnext.config.js%3A4%2C3

### To Reproduce

1. Create any css module file
2. Write a valid native CSS nested code
   ```css
    .Layout {
      border: 10px solid red;
    
      & body {
        height: 100vh;
        margin: 0;
        border: 10px solid blue;
      }
    }

    /* or */

    .Layout {
      border: 10px solid red;
    
      body {
        height: 100vh;
        margin: 0;
        border: 10px solid blue;
      }
    }
   ```
3. See the error about pure selectors

### Current vs. Expected behavior

Nextjs and the check it performs on CSS modules prevents from using the new CSS native nesting syntax (which is now supported by evergreen browsers).

It returns error about a simple nested CSS while this should be allowed as valid CSS syntax now.

Error:

`Selector "& body" is not pure (pure selectors must contain at least one local class or id)`

See the reproduction link.

Next should allow native and valid CSS syntax and avoid extra checks on "pure selectors" where not needed.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
    Binaries:
      Node: 16.17.0
      npm: 8.15.0
      Yarn: 1.22.19
      pnpm: 7.1.0
    Relevant packages:
      next: 13.0.8-canary.0
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Additional context

The selector `.Layout { & body {} }`  should be validated as `.Layout body {}` which is a "pure" css module selector.

**Comment by samir9saeedi** [comment]
FWIW you can use this PostCSS plugin in the meantime: https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-nesting

**Comment by samir9saeedi** [comment]
No worries. The plugin I've linked to (css-nesting) supports the new official syntax, which is slightly different from what SASS uses. That syntax has its own plugin as well (css-nested), which is what you're most likely referring to. Doesn't make a difference in the bundle size presumably, just more future-proof.

**Comment by equinusocio** [comment]
> No worries. The plugin I've linked to (postcss-nesting) supports the new official syntax, which is slightly different from what SASS uses. That syntax has its own plugin as well (postcss-nested), which is what you're most likely referring to. Doesn't make a difference in the bundle size presumably, just more future-proof.

Yep, we use `postcss-nesting` since 2016 via `postcss-preset-env`. Now that the native nesting is available we would like to disable this plugin, we want nested CSS in the browser since the CSS size is reduced drastically.


> Doesn't make a difference in the bundle size presumably, just more future-proof.
Yep, it affects the final CSS drastically, here a basic example:

```css
.MyLongClassName {
  color: red;

  &[data-button-color="blue"] {
    color: blue;
  }
  
  &[data-button-color="green"] {
    color: green;
  }  
}

/* All of this, when processed, become */

.MyLongClassName { color: red }
.MyLongClassName[data-button-color="blue"] { color: blue }
.MyLongClassName[data-button-color="green"] { color: green } 
```

As you can see the parent selector is repeated over and over, now multiply this for every nested selector. Without processing the nested syntax, we just get the exact same code in browsers:

```css
.MyLongClassName {
  color: red;
  &[data-button-color="blue"] { color: blue };
  &[data-button-color="blue"] { color: blue }
} 
```

139 chars minified vs. 108 chars minified. On our side, this is a vast improvement considering the amount of CSS we ship. 😌

**Comment by equinusocio** [comment]
Is this issue planned? Any interest)

**Comment by davidyarham** [comment]
I hope this is planned. Native nesting has been around since March 2023, and this week we have also got relaxed nesting rules natively.

I also agree with @equinusocio, I would rather not have processed CSS when its not needed and let the platform do the work when its supported.

**Comment by landsman** [comment]
I would love to see this. Currently builder blocking me from using them.

**Comment by CamiloGdev** [comment]
> ### Link to the code that reproduces this issue or a replay of the bug
> https://codesandbox.io/p/sandbox/magical-chatelet-9ddsys?file=%2Fnext.config.js%3A4%2C3
> 
> ### To Reproduce
> 1. Create any css module file
> 2. Write a valid native CSS nested code
>    ```css
>     .Layout {
>       border: 10px solid red;
>     
>       & body {
>         height: 100vh;
>         margin: 0;
>         border: 10px solid blue;
>       }
>     }
>    
>     /* or */
>    
>     .Layout {
>       border: 10px solid red;
>     
>       body {
>         height: 100vh;
>         margin: 0;
>         border: 10px solid blue;
>       }
>     }
>    ```
> 3. See the error about pure selectors
> 
> ### Current vs. Expected behavior
> Nextjs and the check it performs on CSS modules prevents from using the new CSS native nesting syntax (which is now supported by evergreen browsers).
> 
> It returns error about a simple nested CSS while this should be allowed as valid CSS syntax now.
> 
> Error:
> 
> `Selector "& body" is not pure (pure selectors must contain at least one local class or id)`
> 
> See the reproduction link.
> 
> Next should allow native and valid CSS syntax and avoid extra checks on "pure selectors" where not needed.
> 
> ### Verify canary release
> * [x]  I verified that the issue exists in the latest Next.js canary release
> 
> ### Provide environment information
> ```shell
> Operating System:
>       Platform: linux
>       Arch: x64
>       Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
>     Binaries:
>       Node: 16.17.0
>       npm: 8.15.0
>       Yarn: 1.22.19
>       pnpm: 7.1.0
>     Relevant packages:
>       next: 13.0.8-canary.0
>       eslint-config-next: N/A
>       react: 18.2.0
>       react-dom: 18.2.0
> ```
> 
> ### Which area(s) are affected? (Select all that apply)
> Not sure
> 
> ### Additional context
> The selector `.Layout { & body {} }` should be validated as `.Layout body {}` which is a "pure" css module selector.

hello my

I had the same problem as you and I'll show you how I solved it:

In this case, nesting only works if the selector to be nested is a class or an id, but if it is the selector of a label it does not work directly, but you can solve this if you name the parent class from which you are nesting again:

>    ```css
>     .Layout {
>       border: 10px solid red;
>     
>       &.layout body {
>         height: 100vh;
>         margin: 0;
>         border: 10px solid blue;
>       }
>     }
>    
>     
>    ```

That is to say that as long as you start with a class or id the nesting and from then on you can call the selectors you want, it is somewhat strange but at least it solves it, but on the other hand I was very happy working this way and then I realized Note that it only works well in client components; I work with Next and it turns out that with client components, nesting CSS in this way works, but as soon as I use it in a server component it stops working, it correctly identifies the classes and their nesting, but it omits properties like flex or some others , this is even rarer, for example

>    ```css
.dataBox {
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	gap: 1rem;
	color: var(--color-S01-800);
	width: 100%;
	border: 1px solid;
	border-color: var(--color-T1-400);
	border-radius: 2rem;
	padding: 4rem;
	
	.dataTitleWrapper {
		display: flex;
		justify-content: space-between;
		align-items: center;
		gap: 1rem;
		width: 100%;
	}
	
	.inlineField {
		width: 100%;
		display: flex;
		gap: 1rem;
		justify-content: center;
		align-items: center;
	}
}

>    ```

In this case, all the properties related to flex in both the .inlineField and .dataTitleWrapper classes, for some reason, when checking with devtools in the browser, they do not appear as if I had not written them and therefore the styles come out wrong, but if I pass them without the nesting, everything it turns out well. but if I pass that same code to a client component everything works out fine.

I DON'T UNDERSTAND ANYTHING

Sorry for my English, I only speak Spanish, it's a pure translator

**Comment by teddarcuri** [comment]
I'd also like to see this feature added/allowed.

**Comment by Leviathan91** [comment]
Since this feature is not included in Nextjs yet, an intermediate solution has been propsed: using postcss-nesting plugin.
I am using NextJs14 with the app directory. While there are [docs on how to integerate postcss into pages router](https://nextjs.org/docs/pages/building-your-application/configuring)
there seem to be no [docs for postcss integration for the app router](https://nextjs.org/docs/app/building-your-application/configuring)
I just tried on a new project with the default settings (npx@latest, spam click enter) and  then adding postcss-nesting and a postcss config file does not work.
What am I missing here? I'd really like to start using the nested css syntax and then lateron remove postcss when the feature is included in canary, instead of having to re-write the css then.

**Comment by equinusocio** [comment]
> Since this feature is not included in Nextjs yet, an intermediate solution has been propsed: using postcss-nesting plugin.
> 
> I am using NextJs14 with the app directory. While there are [docs on how to integerate postcss into pages router](https://nextjs.org/docs/pages/building-your-application/configuring)
> 
> there seem to be no [docs for postcss integration for the app router](https://nextjs.org/docs/app/building-your-application/configuring)
> 
> I just tried on a new project with the default settings (npx@latest, spam click enter) and  then adding postcss-nesting and a postcss config file does not work.
> 
> What am I missing here? I'd really like to start using the nested css syntax and then lateron remove postcss when the feature is included in canary, instead of having to re-write the css then.

Postcss-nesting is not a solution for the issue. It compiles the css by removing the nesting and duplicating the css. Plus by using  browsers list if the target browsers support nesting, the css will stay the same (with nested css) and gives the same error.

**Comment by melodyarcjason** [comment]
Would like to add to this issue - it is making a lot of UI package updates VERY difficult.  In my case - migrating @mantine v6 -> 7 to get away from the CSS-in-JS emotion style so that we are not required to 'use client' on everything.

Not only am I having to convert to the css module - but finding that my nesting is not working, so I have to refactor class names on the component code and flatten my css making the task quite a bit more painful.  An example that I have using css module:

This doesn't apply css to the 'badge':
```
.taskHeader {
  padding: 0rem 1rem;
  display: flex;
  align-items: center;
  gap: 0.5rem;

  opacity: 0.75;
  font-size: 0.75rem;
  font-style: normal;
  font-weight: 600;
  line-height: 1.25rem;

  .mantine-Badge-root{
    height: 1rem;
    min-width: 1.5rem;
    padding: 0rem 0.1rem;
  }
}
```

But the badge is a 'div' so this works oddly enough:
```
.taskHeader {
  padding: 0rem 1rem;
  display: flex;
  align-items: center;
  gap: 0.5rem;

  opacity: 0.75;
  font-size: 0.75rem;
  font-style: normal;
  font-weight: 600;
  line-height: 1.25rem;

  & > div {
    height: 1rem;
    min-width: 1.5rem;
    padding: 0rem 0.1rem;
  }
}
```

As a test I tried using `& > .mantine-Badge-root` and that does not work as a nested selector either.  I even tried `& > div.mantine-Badge-root` and that too fails.

Would be great to get basic css class nesting in place to help adoption and migration efforts.

**Comment by n2k3** [comment]
I'm in the same boat, build error when using CSS modules nested selector.

![image](https://github.com/vercel/next.js/assets/7542020/ab4191e9-18f0-4360-ae1c-c2e8fa69a25f)

Would be nice to see this supported natively in Next.js 👍

**Comment by SSTPIERRE2** [comment]
We can't just forget about core CSS features like this while supporting the bleeding edge of React. The community at large has been anticipating built-in CSS nesting for a long time, and now Next.js devs are still waiting.

**Comment by equinusocio** [comment]
After 5 months no response from them. I guess they don't care.

**Comment by EdwardBock** [comment]
That is very sad because it is a blocker in my agency. If we cannot trust the system to support at least the browsers core features, we cannot use it.

**Comment by legowerewolf** [comment]
Well, damn. I was hoping to switch away from SCSS because all I was using it for was the nesting syntax.

**Comment by acidfernando** [comment]
I'm not sure what nextjs uses to parse the CSS modules, but [css-loader already has fixed this same issue 2 months ago](https://github.com/css-modules/postcss-modules-local-by-default/commit/a8c3d770e18a934892108c2b1903f0cc95cad16d).

Supposing they use css-loader, they just have to update.

**Comment by equinusocio** [comment]
> I'm not sure what nextjs uses to parse the CSS modules, but [css-loader already has fixed this same issue 2 months ago](https://github.com/css-modules/postcss-modules-local-by-default/commit/a8c3d770e18a934892108c2b1903f0cc95cad16d).
> 
> Supposing they use css-loader, they just have to update.

Doesn't seems a fix because CSS nesting doesn't require `&` at the beginning of the selector. The fix they pushed seems to force people to use `&` instead of relying on native CSS syntax. This is a boo boo.

**Comment by ryami333** [comment]
> Supposing they use css-loader, they just have to update.

I can confirm that the PR that you linked to does _not_ address the issue. I just checked my `yarn.lock`, and found that I have version `4.0.4` of `postcss-modules-local-by-default` installed, which includes that change.

**Comment by ryami333** [comment]
> instead of relying on native CSS syntax

Well, the `&` is _also_ native CSS syntax, so it is a _partial_ fix, at least.

**Comment by vicentel89** [comment]
I am using version 14.1.1 and I still get the error.

![image](https://github.com/vercel/next.js/assets/60248266/ce23c2ca-727a-4bc8-a304-b056331e9fdb)
![image](https://github.com/vercel/next.js/assets/60248266/4a4da28b-b653-4668-9943-31448a91e20e)
![image](https://github.com/vercel/next.js/assets/60248266/b512110f-d9eb-4581-a83d-15c8f1bfe874)

**Comment by samcx** [comment]
@vicentel89 Can you try the latest <picture data-single-emoji=":nextjs:" title=":nextjs:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/nextjs/2a85d633ae594556.png" alt=":nextjs:" align="absmiddle"></picture> canary instead?

**Comment by equinusocio** [comment]
> @vicentel89 Can you try the latest <img alt=":nextjs:" width="20" height="auto" src="https://camo.githubusercontent.com/63ff615a24994da53643b9dba36724d41304b284696f2df6cc06e6a32a54890e/68747470733a2f2f656d6f6a692e736c61636b2d656467652e636f6d2f5430434151303054552f6e6578746a732f326138356436333361653539343535362e706e67"> canary instead?

It seems to work, is there any doc update about this? can we use both native syntaxes (with & and without &)?

**Comment by samcx** [comment]
@equinusocio I think this is fine—my current understanding is that CSS **Modules** doesn't support CSS nesting without explicitly stating it → https://github.com/css-modules/postcss-modules-local-by-default/pull/64.

**Comment by equinusocio** [comment]
what happened to this fix? it seems fixed only in `v14.1.1-canary.82` (the only one working). All other versions after this canary don't include the fix anywhere in the changelog or release and are not working. 

Since 14.1.1, 14.1.2, and the latest 14.1.3 are not working the relative issue is not fixed.

**Comment by samcx** [comment]
@equinusocio The past two stable patch versions likely does not have this change (backported stable patch versions). The latest canary does indeed have the dependency bump.
